### PR TITLE
WIP: Api change

### PR DESCRIPTION
**Pull request by mathurinm**
I would expect all Models to behave as sklearn estimators (fit, predict)
In the example is the syntax that I feel is natural : create criterion, create algo, create model, fit model, predict model

Open to discussion !

### CODE REVIEW

File: sparse_ho/models.py [hunk 1]
```diff
@@ -278,20 +292,26 @@ class WeightedLasso():
...
     def __init__(
-            self, X, y, max_iter=1000, estimator=None, log_alpha_max=None):
-        self.X = X
-        self.y = y
+            self, criterion, algo, max_iter=1000, log_alpha_max=None):
+        self.criterion = criterion
+        self.algo = algo
         self.max_iter = max_iter
-        self.estimator = estimator
+        self._estimator = Lasso(warm_start=True)
```
Comment (agramfort) [inline sparse_ho/models.py:304]: this should happen in the fit to be sklearn compliant. The init just set attributes untouched.

So you would do

self.estimator = estimator

here and in the fit

`self.estimator_ = clone(self.estimator) if self.estimator is not None else Lasso(warm_start=True)`

**Comment by QB3** [comment]
Summary of our discussion with @mathurinm :
- we should remove the user estimator instanciation, it should be done automatically. This should be 'easy' for us.
- for the moment the API is thought to solve bilevel optimization problems: we define an inner problem, an outer problem, and an algorithm to solve the bilevel problem:
`my_inner = Inner()
my_outer = Outer(my_inner)
my_algo = Algo(my_outer)
grad_search(my_algo)`

 A machine learning API would be to define the inner problem, and an algo to solve it:
`my_outer = Outer()
my_algo = Algo()
my_inner = Inner(my_outer, my_algo)`

grad_search(my_algo)`

File: sparse_ho/models.py (hunk 1 above)
Comment (mathurinm) [reply sparse_ho/models.py:304]: I would like to avoid passing the estimator in the init
Could I do `self.estimator = None` here ?

**Comment by QB3** [comment]
Another point raised was the addition of a callback function, especially to monitor the loss on a test set of unseen data

**Comment by agramfort** [comment]
I like the idea

When a deeper understanding of the theory helps good design !

>

**Comment by QB3** [comment]
FYI I am putting some link toward API of other hyperparam solvers: 
- hyperopt:  https://hyperopt.github.io/hyperopt/ which takes in argument a function to minimize, and returns ("only"?) the optimal hyperparameter. This is a little bit like sparse-ho today.
```
def objective(args):
    case, val = args
    if case == 'case 1':
        return val
    else:
        return val ** 2

# define a search space
from hyperopt import hp
space = hp.choice('a',
    [
        ('case 1', 1 + hp.lognormal('c1', 0, 1)),
        ('case 2', hp.uniform('c2', -10, 10))
    ])

# minimize the objective over the space
from hyperopt import fmin, tpe
best = fmin(objective, space, algo=tpe.suggest, max_evals=100)
```
- hyperopt-sklearn: https://hyperopt.github.io/hyperopt-sklearn/, which seems to "overwrite" the sklearn estimators, which is more user friendly.:
```
from hpsklearn import HyperoptEstimator

# Load Data
# ...

# Create the estimator object
estim = HyperoptEstimator()

# Search the space of classifiers and preprocessing steps and their
# respective hyperparameters in sklearn to fit a model to the data
estim.fit( train_data, train_label )

# Make a prediction using the optimized model
prediction = estim.predict( unknown_data )

# Report the accuracy of the classifier on a given set of data
score = estim.score( test_data, test_label )

# Return instances of the classifier and preprocessing steps
model = estim.best_model()```

**Comment by mathurinm** [comment]
We split into smaller PRs (trottinette style)
- separate criterion from algo
- separate model from criterion
- make model behave like a sklearn estimator

**Comment by svaiter** [comment]
> FYI I am putting some link toward API of other hyperparam solvers:
> 
Other libraries of interest includes:

- https://scikit-optimize.github.io/stable/
- https://optuna.org
- https://docs.ray.io/en/latest/tune/index.html

I don't think we need to cover all existing hyper-parameters tuning strategies in the world, but the three mentioned have >2k stars, so you we could consider them. Note: I never used either of three.
